### ISSUE TITLE
Regularly get `not enough bytes`

### ISSUE THREAD

**Issue by teto** [erebe/greenclip#80]
I love how greenclip integrates with rofi-greenclip but it's hardly usable because every couple minutes I end up clearing the cache because of https://github.com/erebe/greenclip/issues/31

see
```
13:03:17 jedha greenclip[18163]:   error, called at libraries/binary/src/Data/Binary/Get.hs:351:5 in binary-0.8.7.0:Dat>
mars 29 13:03:17 jedha greenclip[18163]: Data.Binary.Get.runGet at position 0: not enough bytes
mars 29 13:03:17 jedha greenclip[18163]: CallStack (from HasCallStack):
mars 29 13:03:17 jedha greenclip[18163]:   error, called at libraries/binary/src/Data/Binary/Get.hs:351:5 in binary-0.8.7.0:Dat>
mars 29 13:03:17 jedha greenclip[18163]: Data.Binary.Get.runGet at position 0: not enough bytes
mars 29 13:03:17 jedha greenclip[18163]: CallStack (from HasCallStack):
```
Could it be more robust ?

**Comment by erebe** [comment]
Hello,
Greenclip has been pretty stable for a long time and nobody brought up issues for a long time.
I use it also daily without any issues. All of that to say we need to understand what is different on your setup.

1. Are you sure you don't have 2 greenclip daemon running at the same time ? (which is the most probable issue)
2. What is your setup, Window Manager, locale, ...
3. Can you send the corrupted history file

**Comment by teto** [comment]
I am using i3 on nixos, greenclip is run as a user service:
```
➜ systemctl --user cat greenclip                                                                                     ~/dotfiles
# /nix/store/shcpsgkz93cvg9jfqag5y2x0nrcn7fba-unit-greenclip.service/greenclip.service
[Unit]
After=graphical-session.target
Description=greenclip daemon

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/k51wcq4lqkj3la51jlpnsnnly5wwhc8i-glibc-locales-2.30/lib/locale/locale-archive"
Environment="PATH=/nix/store/hzvl3dvv8651iqlb5g6gq5hzjzmhjn7m-coreutils-8.31/bin:/nix/store/rphcpivxfm4blw36ki262yrmqygw9pcp-fi>
Environment="TZDIR=/nix/store/msbzkf26r6v5nypj1dspjk5jfnvl9l8y-tzdata-2019c/share/zoneinfo"



ExecStart=/nix/store/xw62hz5k0ksb9k03h02ksn4swkd36c8g-greenclip-3.2.0/bin/greenclip daemon
```
My locale (with accents,
```
➜ locale                                                                                                             ~/dotfiles
LANG=fr_FR.UTF-8
LC_CTYPE="fr_FR.UTF-8"
LC_NUMERIC="fr_FR.UTF-8"
LC_TIME="fr_FR.UTF-8"
LC_COLLATE="fr_FR.UTF-8"
LC_MONETARY="fr_FR.UTF-8"
LC_MESSAGES="fr_FR.UTF-8"
LC_PAPER="fr_FR.UTF-8"
LC_NAME="fr_FR.UTF-8"
LC_ADDRESS="fr_FR.UTF-8"
LC_TELEPHONE="fr_FR.UTF-8"
LC_MEASUREMENT="fr_FR.UTF-8"
LC_IDENTIFICATION="fr_FR.UTF-8"
LC_ALL=
```

my config
```
Config {
 maxHistoryLength = 50,
 historyPath = "~/.cache/greenclip.history",
 staticHistoryPath = "~/.cache/greenclip.staticHistory",
 imageCachePath = "/tmp/",
 usePrimarySelectionAsInput = True,
 blacklistedApps = [],
 trimSpaceFromSelection = True
}
```

I will add the history file once it gets corrupt (shouldn't take too long xD).

**Comment by teto** [comment]
It appears that the file is empty when I get the error.
I think I know when it happens, I noticed it usually happens a reconfigure of my desktop (I use nixos), which restarts greenclip. For some reason the restart seems to confuse greenclip and make it lose its way in the file.

I wish `greenclip print' -show clipboard'` would tell me about such errors or fail. Right now I have to try to paste things to see if anything happened.

I also had this failure this morning.
```
avril 07 13:44:56 jedha systemd[1437]: Started greenclip daemon.
avril 07 14:41:12 jedha greenclip[1512]: X Error of failed request:  BadWindow (invalid Window parameter)
avril 07 14:41:12 jedha greenclip[1512]:   Major opcode of failed request:  20 (X_GetProperty)
avril 07 14:41:12 jedha greenclip[1512]:   Resource id in failed request:  0x5c00001
avril 07 14:41:12 jedha greenclip[1512]:   Serial number of failed request:  86621
avril 07 14:41:12 jedha greenclip[1512]:   Current serial number in output stream:  86621
avril 07 14:41:12 jedha systemd[1437]: greenclip.service: Main process exited, code=exited, status=1/FAILURE
```

**Comment by erebe** [comment]
Hello,

greenclip print' -show clipboard' is only reading the history file from disk, so there is no way for it to tell that X is broken, as it does not even create an X window.

For your errors, I don't know what NixOS is doing but it is rather nasty. Basically greenclip is trying to request information from a window that is no longer registered with Xorg. This error cannot be caught, it is Xorg that is killing greenclip for violating the X protocol, as its window is not registered.

I suppose your Xorg is being restarted while greenclip daemon is not, so on next request greenclip get killed as Xorg does not know about it.

That being said, I have been a user of NixOS in the past and would rather recommend you not using it as your OS for your Desktop. It is great for CI and stuff but will all the "hack" put in place for it to work you will stumble, like me, on a lot of edge cases with normal software.

**Comment by teto** [comment]
```
I suppose your Xorg is being restarted while greenclip daemon is not, so on next request greenclip get killed as Xorg does not know about it.
```
Yeah, the greenclip service is handled by the system instead of home-manager, sthg I certainly need to fix (as well as the systemd chain maybe).

This was the first time I saw this error but I wonder if there couldn't be some kind of check run by greenclip on startup since when I have an issue, even restarting greenclip won't solve it, I really need to delete the history file.

As for nixos, I know it can be tricky (I contribute a lot) but I find it a real blessing, it has made my computer experience so much more reliable. Still rough edges sadly.

**Comment by erebe** [comment]
Going to look at it to prevent manual intervention for not enough bytes

**Comment by erebe** [comment]
Should be fixed in release 3.3, the history file will be emptied if corrupted. So no need for a manual intervention.

The issue was that `decode` was throwing an uncatchable exception ...

Archlinux package and Hackage have been updated to 3.3 also

**Comment by teto** [comment]
fantastic thanks !
